docs/languages/go.md

```diff
@@ -402,3 +402,249 @@ func Handle(w http.ResponseWriter, r *http.Request) {
+		}
+	}
+
+	return exporters, nil
+}
+```
+
+The `newTracerProvider` initializes a new tracer provider using the provided exporters.
+
+```go
+func newTracerProvider(exporters []sdktrace.SpanExporter) *sdktrace.TracerProvider {
+	serviceName := os.Getenv("OTEL_SERVICE_NAME")
+	if serviceName == "" {
+		serviceName = "golang-middleware-function"
+	}
+
+	// Ensure default SDK resources and the required service name are set.
+	r, err := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceName(serviceName),
+		),
+	)
+
+	if err != nil {
+		panic(err)
+	}
+
+	opts := []sdktrace.TracerProviderOption{
+		sdktrace.WithResource(r),
+	}
+
+	for _, exporter := range exporters {
+		processor := sdktrace.NewBatchSpanProcessor(exporter)
+		opts = append(opts, sdktrace.WithSpanProcessor(processor))
+	}
+
+	return sdktrace.NewTracerProvider(opts...)
+}
+```
+
+Use the modified template to create a new function.
+
+```sh
+faas-cli new echo --lang golang-middleware-otel
+```
+
+Use environment variables to configure the traces exporter for the function.
+
+```diff
+functions:
+  greet:
+    lang: python3-http-otel
+    handler: ./greet
+    image: greet:latest
++    environment:
++      OTEL_SERVICE_NAME: greet.${NAMESPACE:-openfaas-fn}
++      OTEL_TRACES_EXPORTER: console,otlp
++      OTEL_EXPORTER_OTLP_ENDPOINT: ${OTEL_EXPORTER_OTLP_ENDPOINT:-collector:4317}
+```
+
+- `OTEL_SERVICE_NAME` sets the name of the service associated with the telemetry and is used to identify telemetry for a specific function. It can be set to any value you want but we recommend using the clear function identifier `<fn-name>.<fn-namespace>`.
+- `OTEL_TRACES_EXPORTER` specifies which tracer exporter to use. In this example traces are exported to `console` (stdout) and with `otlp` to send traces to an endpoint that accepts OTLP via gRPC.
+- `OTEL_EXPORTER_OTLP_ENDPOINT` sets the endpoint where telemetry is exported to.
+
+### Creating custom traces in the function handler
+
+When using a modified OpenTelemetry template the global trace provider can be retrieved to add custom spans or additional instrumentation libraries in the function handler.
+
+Make sure the install the required packages:
+
+```sh
+go get "go.opentelemetry.io/otel" \
+	"go.opentelemetry.io/otel/trace"
+```
+
+Add custom spans in the function handler:
+
+```go
+package function
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"go.opentelemetry.io/otel"
+)
+
+var tracer trace.Tracer
+var mu sync.Mutex
+
+func getTracer() trace.Tracer {
+	if tracer == nil {
+		mu.Lock()
+		tracer = otel.GetTracerProvider().Tracer("function")
+		mu.Unlock()
+	}
+
+	return tracer
+}
+
+
+func Handle(w http.ResponseWriter, r *http.Request) {
+	ctx, span := getTracer().Start(r.Context(), "function-span")
+	defer span.End()
+
+	var input []byte
+
+	if r.Body != nil {
+		defer r.Body.Close()
+
+		body, _ := io.ReadAll(r.Body)
+
+		input = body
+	}
+
```

Comment:
Can you create a method i.e. "callOpenAI" - make it wait 2 seconds, create a separate span for it?